src/details/ArborX_DetailsBufferOptimization.hpp

```diff
@@ -0,0 +1,376 @@
+  }
+};
+
+} // namespace Details
+
+namespace Traits
+{
+template <typename Predicates, typename Permute>
+struct Access<Details::PermutedPredicates<Predicates, Permute>, PredicatesTag>
+{
+  using PermutedPredicates = Details::PermutedPredicates<Predicates, Permute>;
+  using NativeAccess = Access<Predicates, PredicatesTag>;
+
+  inline static std::size_t size(PermutedPredicates const &permuted_predicates)
+  {
+    return NativeAccess::size(permuted_predicates._predicates);
+  }
+
+  KOKKOS_INLINE_FUNCTION static auto
+  get(PermutedPredicates const &permuted_predicates, std::size_t i)
+  {
+    return NativeAccess::get(permuted_predicates._predicates,
+                             permuted_predicates._permute(i));
+  }
+  using memory_space = typename NativeAccess::memory_space;
+};
+} // namespace Traits
+
+namespace Details
+{
+
+template <typename ExecutionSpace, typename TreeTraversal, typename Predicates,
+          typename Callback, typename OutputView, typename OffsetView,
+          typename PermuteType>
+void queryImpl(ExecutionSpace const &space, TreeTraversal const &tree_traversal,
+               Predicates const &predicates, Callback const &callback,
+               OutputView &out, OffsetView &offset, PermuteType permute,
+               BufferStatus buffer_status)
+{
+  // pre-condition: offset and out are preallocated. If buffer_size > 0, offset
+  // is pre-initialized
+
+  static_assert(Kokkos::is_execution_space<ExecutionSpace>{}, "");
+
+  using Access = Traits::Access<Predicates, Traits::PredicatesTag>;
+  auto const n_queries = Access::size(predicates);
+
+  Kokkos::Profiling::pushRegion("ArborX:BVH:two_pass");
+
+  using CountView = OffsetView;
+  CountView counts(Kokkos::view_alloc("counts", space), n_queries);
+
+  using PermutedPredicates = PermutedPredicates<Predicates, PermuteType>;
+  PermutedPredicates permuted_predicates = {predicates, permute};
+
+  Kokkos::Profiling::pushRegion("ArborX:BVH:two_pass:first_pass");
+  bool underflow = false;
+  bool overflow = false;
+  if (buffer_status != BufferStatus::PreallocationNone)
+  {
+    tree_traversal.launch(
+        space, permuted_predicates,
+        InsertGenerator<FirstPassTag, PermutedPredicates, Callback, OutputView,
+                        CountView, OffsetView, PermuteType>{
+            permuted_predicates, callback, out, counts, offset, permute});
+
+    // Detecting overflow is a local operation that needs to be done for every
+    // index. We allow individual buffer sizes to differ, so it's not as easy
+    // as computing max counts.
+    int overflow_int = 0;
+    Kokkos::parallel_reduce(
+        ARBORX_MARK_REGION("compute_overflow"),
+        Kokkos::RangePolicy<ExecutionSpace>(space, 0, n_queries),
+        KOKKOS_LAMBDA(int i, int &update) {
+          auto const *const offset_ptr = &offset(permute(i));
+          if (counts(i) > *(offset_ptr + 1) - *offset_ptr)
+            update = 1;
+        },
+        overflow_int);
```

Comment:
The whole overflow calculation is trivial for the case of constant buffer per entry. Right now, it is in generic form.